opentelemetry-sdk/src/logs/record.rs

```diff
@@ -124,6 +125,73 @@ impl LogRecord {
             .flatten()
             .any(|(k, v)| k == key && v == value)
     }
+
+    /// Updates the first occurrence of an attribute with the specified key.
+    ///
+    /// This method searches for the first occurrence of the attribute with the given key
+    /// in the `attributes` collection. If the key is found, its value is updated with the
+    /// provided value. If the key is not found, the attribute is added.
+    ///
+    /// # Arguments
+    ///
+    /// - `key`: A reference to the key of the attribute to update.
+    /// - `value`: A new value for the attribute.
+    ///
+    /// # Returns
+    ///
+    /// - `Some(AnyValue)`: The old value of the attribute if found and updated.
+    /// - `None`: If the attribute was not found, and a new one was added.
+    ///
+    pub fn update_attribute(&mut self, key: &Key, value: AnyValue) -> Option<AnyValue> {
+        // First, search for the attribute mutably
+        if let Some(attr) = self
+            .attributes
+            .iter_mut()
+            .find(|opt| opt.as_ref().map(|(k, _)| k == key).unwrap_or(false))
+        {
+            // Take the old value and update the attribute
+            let old_value = attr.take().map(|(_, v)| v);
+            *attr = Some((key.clone(), value));
+            return old_value;
+        }
+
+        // If not found, add a new attribute
+        self.add_attribute(key.clone(), value.clone());
+        None
```

Comment:
Small nit on code style

maybe easier to read if we do 


```rust
match self
            .attributes
            .iter_mut()
            .find(|opt| opt.as_ref().map(|(k, _)| k == key).unwrap_or(false)) {
                     Some(attr) => {
                               // Take the old value and update the attribute
                               let old_value = attr.take().map(|(_, v)| v);
                               *attr = Some((key.clone(), value));
                               old_value
                        }
                     None => self.add_attribute(key.clone(), value.clone());
}
```

Reply:
I was little confused why we need to `add_attribute` regardless until I see the early return

Reply:
Or you have already tried that and compiler complains because the we are still holding mutable reference to self when iterator found nothing?

Reply:
> Or you have already tried that and compiler complains because the we are still holding mutable reference to self when iterator found nothing?

@TommyCpp yes, i did try that, and got the error for still holding the mutable reference to self. Looks like the borrow checker is bit conservative and doesn't automatically drop iterators/temporaries until the end of the scope where they are used.